### PR TITLE
Optimize allocations in LogRotator and RunList

### PR DESCRIPTION
**Pull request by gulyaev13**
LogRotator and RunList allocated a lot of arrays during the creation of the sub-list.
Fix contains small optimizations.

### Proposed changelog entries

* Entry 1: Set initial capacity in RunList.subList
* Entry 2: Don't copy already a new instance of List<Run> in LogRotator.perform

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by oleg-nenashev — APPROVED** [review]
It is also documented in Javadoc: https://javadoc.jenkins-ci.org/hudson/util/RunList.html#subList-int-int- . We will need to ensure that the behavior is preserved in the future.

**Review by res0nance — APPROVED** [review]
Removing the copy is fine it was added in #573 8 years ago because it used to use AbstractList's sublist implementation which was not a copy but later a sublist implementation was added in 807dc7174cd8ed4ec3cb9562f1c3f90f4035cf33 which simply made a copy of the list. This was probably due to iterator invalidation due to Run.delete() removing the run from the internal map. This should be fine

**Review by timja — APPROVED** [review]

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**
